performant-software/DM
# Text format safeguarding

### What does this PR do?
When their contents have changed, text editors will check for the kind of style map attached to blocks of rich text when they are pasted in. If found (if text has just been pasted in), the editor will convert the generic class names used in the style maps into unique names. This will prevent the collision of repeated class names in rich text blocks pasted in at different times or in different documents.

### What issues does it address?
Closes #68 (removes need for nuclear option), closes #138 

### How to test
- At http://dm-staging.middell.net/, open or create a project with a text document. Enter some text manually, and also paste in text from a variety of sources, namely different text editors. Adding new text should never affect the formatting of text that was already present in the document (newly added text may take on unintended styles from text already present, e.g. pasting a line of black text right after some red text that was already there will probably result in both lines being red, but the reverse should never happen - pasting in red text after a line of black text that was already there should never turn the preexisting text red). Try to throw a large amount of differently formatted text copied from different text editors together in the document.
-  In a separate text editor, create a few lines of text with different formatting. Copy them into a text resource in DM and put the resource in read-only mode. Change the styling on a line or two of the original text in your text editor, then copy this to a new resource and put it too in read-only mode. With both resources open in read-only mode, ensure that the styling appears in both as it did when you pasted it in. Try dragging the resources into a different order in the grid and opening more text resources and ensure the styling doesn't change.
